### PR TITLE
Windows Core Build Support

### PR DESCRIPTION
**Pull request by nibanks**
This PR builds off the changes in #579 and adds Windows build support and AppVeyor CI build support.

### CODE REVIEW

File: src/core/mac/mac.cpp [hunk 1]
```diff
@@ -269,7 +269,11 @@ ThreadError Mac::SetNetworkName(const char *aNetworkName)
     VerifyOrExit(strlen(aNetworkName) <= OT_NETWORK_NAME_MAX_SIZE, error = kThreadError_InvalidArgs);
 
     memset(&mNetworkName, 0, sizeof(mNetworkName));
+#ifdef _WIN32
+    strcpy_s(mNetworkName.m8, sizeof(mNetworkName), aNetworkName);
+#else
```
Comment (darconeous) [inline src/core/mac/mac.cpp]: If `strncpy()` isn't present on the platform, we should provide a replacement function for it (or even use a preprocessor macro). Let's not add `_WIN32` checks everywhere like this, it hurts readability.

File: src/core/openthread.cpp [hunk 2]
```diff
@@ -864,11 +864,15 @@ void otSetStateChangedCallback(otInstance *, otStateChangedCallback aCallback, v
...
 {
     static const char sVersion[] =
-        PACKAGE_NAME "/" PACKAGE_VERSION "; "
+        PACKAGE_NAME "/" PACKAGE_VERSION
 #ifdef  PLATFORM_INFO
-        PLATFORM_INFO "; "
+        "; " PLATFORM_INFO
+#endif
+#if !defined(WINDOWS_KERNEL) || defined(DBG)
+        "; " __DATE__ " " __TIME__;
```
Comment (darconeous) [inline src/core/openthread.cpp]: What's the problem here with `__DATE__` and `__TIME__`?

File: src/core/thread/network_data.cpp [hunk 3]
```diff
@@ -592,7 +592,7 @@ ThreadError NetworkData::SendServerDataNotification(bool aLocal, uint16_t aRloc1
     Message *message;
     Ip6::MessageInfo messageInfo;
 
-    mSocket.Open(&HandleUdpReceive, this);
+    mSocket.Open(&NetworkData::HandleUdpReceive, this);
```
Comment (darconeous) [inline src/core/thread/network_data.cpp]: Aren't symbols like this already in scope? Why do we need to be explicit here?

File: src/core/mac/mac.cpp (hunk 1 above)
Comment (nibanks) [inline src/core/mac/mac.cpp]: Windows throws a warning/error because it is deprecated (unsecure). See the [MSDN docs](https://msdn.microsoft.com/en-us/library/xdsywd25.aspx) for more info.

File: src/core/openthread.cpp (hunk 2 above)
Comment (nibanks) [inline src/core/openthread.cpp]: They aren't defined in Windows kernel mode in RELEASE mode. I don't know why.

File: src/core/thread/network_data.cpp (hunk 3 above)
Comment (nibanks) [inline src/core/thread/network_data.cpp]: MSBUILD throws an error. It seems to support one of the following patterns (openthread uses both):

1. `mSocket.Open(HandleUdpReceive, this);`
2. `mSocket.Open(&NetworkData::HandleUdpReceive, this);`

Ideally, we should use the same syntax everywhere, but I'm not necessarily signing up for that. :)

File: third_party/mbedtls/repo/include/mbedtls/aes.h [hunk 4]
```diff
@@ -30,7 +30,7 @@
 #endif
 
 #include <stddef.h>
-#include <stdint.h>
+#include <openthread-std-types.h>
```
Comment (darconeous) [inline third_party/mbedtls/repo/include/mbedtls/aes.h]: These changes (changing all references to `<stdint.h>` to `<openthread-std-types.h>`) are not acceptable.

An acceptable alternative would be to create a file named `stdint.h` in the folder `src/replacements/stdint/`. This folder would be included in the system path on systems without `stdint.h`.

File: third_party/mbedtls/repo/library/ctr_drbg.c [hunk 5]
```diff
@@ -178,7 +178,7 @@ static int block_cipher_df( unsigned char *output,
     buf_len = MBEDTLS_CTR_DRBG_BLOCKSIZE + 8 + data_len + 1;
 
     for( i = 0; i < MBEDTLS_CTR_DRBG_KEYSIZE; i++ )
-        key[i] = i;
+        key[i] = (unsigned char)i;
```
Comment (darconeous) [inline third_party/mbedtls/repo/library/ctr_drbg.c]: Changes like this should be contributed upstream. We generally want to keep all sources in `third_party` as pristine releases from those projects, but exceptions are made on a temporary basis assuming the changes have been contributed upstream and `README.google` has been updated to indicate what changes have been made.

File: appveyor.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+version: 0.1.{build}
```
Comment (darconeous) [inline appveyor.yml]: Can this file be renamed to `.appveyor.yml` (So that it is hidden on unixy platforms), or moved to a different directory? We would rather not clutter up the root of this project.

File: examples/platforms/cc2538/alarm.c [hunk 7]
```diff
@@ -32,9 +32,6 @@
  *
  */
 
-#include <stdbool.h>
```
Comment (darconeous) [inline examples/platforms/cc2538/alarm.c]: Why the changes to the cc2538 files? Windows targets shouldn't be building these files.

File: include/openthread-types.h [hunk 8]
```diff
@@ -243,7 +241,11 @@ typedef struct otExtAddress
 /**
  * This structure represents an IPv6 address.
  */
+#ifdef _WIN32
+OT_TOOL_PACKED_BEGIN typedef struct otIp6Address
```
Comment (darconeous) [inline include/openthread-types.h]: There has got to be a better way. Perhaps if we define the struct first, and then define the typedef afterward?

File: include/openthread-windows-config.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#define PACKAGE_STRING "OPENTHREAD 0.01.00"
+
+/* Define to the one symbol short name of this package. */
+#define PACKAGE_TARNAME "openthread"
+
+/* Define to the home page for this package. */
+#define PACKAGE_URL "http://github.com/openthread/openthread"
+
+/* Define to the version of this package. */
+#define PACKAGE_VERSION "0.01.00"
```
Comment (darconeous) [inline include/openthread-windows-config.h]: Maintaining this is going to be difficult long-term. Is there any way to have visual studio create this file as a part of the build process, perhaps using PowerShell? The current version can always be found in the file named `.default-version` in the root of the project.

File: appveyor.yml (hunk 6 above)
Comment (nibanks) [inline appveyor.yml]: I don't think so, according to appveyor's [docs](https://www.appveyor.com/docs/build-configuration/#appveyoryml).

File: examples/platforms/cc2538/alarm.c (hunk 7 above)
Comment (nibanks) [inline examples/platforms/cc2538/alarm.c]: I might of got a little carried away with the find/replace... BTW, I think it would be better to make comments related to the std*.h file replacements in the PR specifically for that: #579.

File: openthread.sln [hunk 10]
```diff
@@ -0,0 +1,79 @@
+﻿
```
Comment (darconeous) [inline openthread.sln]: We would prefer all non-autotools build/project files be located in a subdirectory of `etc`: In this case, perhaps `etc/visual-studio/`. This goes for all related Visual Studio project files.

File: src/core/common/message.hpp [hunk 11]
```diff
@@ -34,14 +34,17 @@
...
-#include <stdint.h>
 #include <string.h>
 
 #include <openthread-types.h>
 #include <openthread-core-config.h>
 #include <common/code_utils.hpp>
 #include <mac/mac_frame.hpp>
 
+#ifdef _WIN32
+#pragma warning(disable:4201)  //nonstandard extension used : nameless struct/union
```
Comment (darconeous) [inline src/core/common/message.hpp]: Visual Studio C++ doesn't support nameless unions...?

Comment (nibanks) [inline src/core/common/message.hpp]: It supports it, but again, in kernel mode, it throws a warning. I don't know why.

File: src/core/thread/meshcop_tlvs.hpp [hunk 12]
```diff
@@ -1329,7 +1330,7 @@ class DiscoveryRequestTlv: public Tlv
      * @retval FALSE  If the Joiner flag is not set.
      *
      */
-    bool IsJoiner(void) { return mFlags & kJoinerMask; }
+    bool IsJoiner(void) { return (mFlags & kJoinerMask) != 0; }
```
Comment (darconeous) [inline src/core/thread/meshcop_tlvs.hpp]: Good catch. 👍

Comment (nibanks) [inline src/core/thread/meshcop_tlvs.hpp]: Again, a Windows warning... I guess they can be good for something.

File: src/core/mac/mac.cpp (hunk 1 above)
Comment (darconeous) [inline src/core/mac/mac.cpp]: Agreed about `strncpy()`. Indeed, the given usage is actually incorrect! Not sure why I didn't notice earlier.

I think the appropriate replacement function would be [`strlcpy`](https://www.freebsd.org/cgi/man.cgi?query=strlcpy&sektion=3), which works as everyone expects it should. It is available on BSD platforms, and its behavior is easily replicated in a replacement implementation.

That will fix the bug on all platforms.

File: src/core/openthread.cpp (hunk 2 above)
Comment (darconeous) [inline src/core/openthread.cpp]: How about we make this `#ifdef __DATE__` then?

File: src/core/thread/network_data.cpp (hunk 3 above)
Comment (darconeous) [inline src/core/thread/network_data.cpp]: Well, I'm OK with the more-specific scope. Was just curious what prompted it.

File: third_party/mbedtls/repo/include/mbedtls/aes.h (hunk 4 above)
Comment (nibanks) [inline third_party/mbedtls/repo/include/mbedtls/aes.h]: Ok. `src/replacements/stdint/stdint.h` is the exact place you want the file?

File: appveyor.yml (hunk 6 above)
Comment (darconeous) [inline appveyor.yml]: Filed [appveyor/ci/#1040](https://github.com/appveyor/ci/issues/1040). We'll see what happens.

File: examples/platforms/cc2538/alarm.c (hunk 7 above)
Comment (darconeous) [inline examples/platforms/cc2538/alarm.c]: Will transplant my comments over there.

File: third_party/mbedtls/repo/include/mbedtls/aes.h (hunk 4 above)
Comment (darconeous) [inline third_party/mbedtls/repo/include/mbedtls/aes.h]: Discussing in #579

File: src/core/mac/mac.cpp (hunk 1 above)
Comment (darconeous) [inline src/core/mac/mac.cpp]: Thinking more about the path, I think I like `src/missing/*` better as the path, since that matches up better with other open source projects.

Comment (darconeous) [inline src/core/mac/mac.cpp]: This could likely be spun off into it's own pull request, considering that it affects all platforms.

File: src/core/common/logging.cpp [hunk 13]
```diff
@@ -31,10 +31,17 @@
...
+#include OPENTHREAD_CONFIG_FILE
+#else
+#include <openthread-config.h>
+#endif
+
 #include <ctype.h>
 #include <stdio.h>
 #include <string.h>
 
+#include <common/code_utils.hpp>
```
Comment (nibanks) [inline src/core/common/logging.cpp]: Oops. Didn't mean to add this.

File: src/core/common/new.hpp [hunk 14]
```diff
@@ -36,6 +36,10 @@
 
 #include <stddef.h>
 
+#ifdef _WIN32
+#pragma warning(disable:4291)  // no matching operator delete found
+#endif
+
```
Comment (nibanks) [inline src/core/common/new.hpp]: Going to move this to the config file as well.

File: openthread.sln (hunk 10 above)
Comment (nibanks) [inline openthread.sln]: Would you mind if I left the solution file at the root? It greatly simplifies the output directory copying and linking between projects.

File: src/core/mac/mac.cpp (hunk 1 above)
Comment (nibanks) [inline src/core/mac/mac.cpp]: `strlcpy` doesn't seem to work. It doesn't exist on Windows and when I compile on Linux:
```
mac/mac.cpp: In member function ‘ThreadError Thread::Mac::Mac::SetNetworkName(const char*)’:
mac/mac.cpp:272:64: error: ‘strlcpy’ was not declared in this scope
     strlcpy(mNetworkName.m8, aNetworkName, sizeof(mNetworkName));
```

File: third_party/mbedtls/repo/library/ctr_drbg.c (hunk 5 above)
Comment (nibanks) [inline third_party/mbedtls/repo/library/ctr_drbg.c]: Where would I contribute the fixes to?

File: appveyor.yml (hunk 6 above)
Comment (darconeous) [inline appveyor.yml]: IlyaFinkelshteyn commented:

> Hi Robert,
> 
> We just recently added support for `.appveyor.yml` with `.` Please give it a try. Note that > appveyor.yml without `.` Is still a first choice.

🎉 

So let's go with `.appveyor.yml`.

**Comment by darconeous** [comment]
After reading (and making a first past at a response to) [this comment](https://github.com/openthread/openthread/pull/582#discussion_r78650136), some of the long-term scope of this change started to dawn on me.

Currently, Windows is not an officially supported platform.  Were it not for you sending us your carefully-crafted pull requests (that happen to also clean up some of our engineering debt), Windows wouldn't have even been on our radar.

To the extent that adding unofficial Windows support helps push our code base into a better, more flexible, and more easily maintained state, that's great. So far, that has been the case.

With this commit, things become at least semi-official. If it's in the code base, it needs to be maintained, curated, and occasionally fixed.

I'm wondering two things:

* Is there a way that we can make the maintenance of the visual studio files automatic? As in, somehow derived from the autoconf makefiles? (I did something like this 15 years ago for a school project, but I'm sure things have changed a lot since then)
* How committed are you (and/or your employer) to maintaining Windows support long term?

**Comment by nibanks** [comment]
As far as I could find, there is no easy solution to substitute what you have for UNIX systems on Windows. I had to manually put together the VS project files to get where I am today. I can look a bit more to see if I can find any solution; perhaps even putting together a script/program myself that is a partial solution.

As far as committed, at least for the time being and the foreseeable future (I cannot make guarentees forever) Microsoft is committed to getting a Windows solution to Thread, and we are invested in OpenThread as that solution. I am personally working on this most of my time, and we have one other dev working on commissioning / border router specific stuff to Windows as well.

File: third_party/mbedtls/repo/library/ctr_drbg.c (hunk 5 above)
Comment (darconeous) [inline third_party/mbedtls/repo/library/ctr_drbg.c]: Looking at the [`README.google`](https://github.com/openthread/openthread/blob/master/third_party/mbedtls/README.google) file, it looks like https://github.com/ARMmbed/mbedtls

**Comment by darconeous** [comment]
I wouldn't stress out too much about trying to set up automatic Visual Studio file generation. It isn't critical.

After merging this commit we don't intend to make Appveyor-CI a mandatory pass for commits. However, we will try to include it for advisory purposes. To the extent that we can easily prevent Appveyor-CI from going red, we will, but if it goes red anyway then it will ultimately be your (employer's) responsibility to address it. As long as we have that understanding then I think we will be golden.

We are discussing the "solution file" location. It isn't clear exactly how much more difficult it would be for it to be located off-root. I take it is more than just prepending `..\..\` to all of the paths...?

**Comment by nibanks** [comment]
If just the project files are in the separate folder, then only the path to the source files needs to be updated from what I have already. The output binaries and input linkages are relative to $(SolutionPath), so if that changes also, then all the output binary & object paths and input linking paths have to change as well.

**Comment by darconeous** [comment]
Couldn't you just have a `sed` script do a find-and-replace for all of the paths? Something like this:

```
find . -name '*.vcxproj' | xargs -n 1 sed -i "" -e 's/\$(SolutionDir)\([^.][^.]\)/\$(SolutionDir)..\\..\\\1/g'
```

I don't have visual studio to test this change, but I've gone ahead and applied it on my own branch as f326d904ccb23c56a7dc5f3fbdc6eff1f9bc65dd. Is there something else I'm missing?

**Comment by nibanks** [comment]
No, it's not really a matter of difficulty, but just uglier to use and maintain. When using the UI, you never see the file paths, but you do see and use and update the other paths. Making them relative to something other than the root will make it confusing. If you really don't want the one solution file at the root, I will make the changes, but I would prefer to keep it at the root.

File: include/openthread-windows-config.h [hunk 15]
```diff
@@ -0,0 +1,78 @@
+/*
+ *  Copyright (c) 2016, Microsoft Corporation.
```
Comment (nibanks) [inline include/openthread-windows-config.h]: Update to use OpenThread Authors

File: include/openthread-windows-config.h (hunk 9 above)
Comment (nibanks) [inline include/openthread-windows-config.h]: For the time being, I will commit to keeping it up to date. I will look into automated solutions.

**Comment by darconeous** [comment]
@jwhui, what are your thoughts on that?

**Comment by nibanks** [comment]
I am seeing this failure all the time now. Can we maybe disable this test?

```
======================================================================
ERROR: test (__main__.Cert_9_2_14_PanIdQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/openthread/openthread/examples/../tests/scripts/thread-cert/Cert_9_2_14_PanIdQuery.py", line 95, in test
    self.nodes[COMMISSIONER].panid_query('0xdead', '0xffffffff', ipaddr)
  File "/home/travis/build/openthread/openthread/tests/scripts/thread-cert/node.py", line 328, in panid_query
    self.pexpect.expect('Conflict:', timeout=8)
  File "/usr/local/lib/python2.7/dist-packages/pexpect/spawnbase.py", line 321, in expect
    timeout, searchwindowsize, async)
  File "/usr/local/lib/python2.7/dist-packages/pexpect/spawnbase.py", line 345, in expect_list
    return exp.expect_loop(timeout)
  File "/usr/local/lib/python2.7/dist-packages/pexpect/expect.py", line 107, in expect_loop
    return self.timeout(e)
  File "/usr/local/lib/python2.7/dist-packages/pexpect/expect.py", line 70, in timeout
    raise TIMEOUT(msg)
TIMEOUT: Timeout exceeded.
```

**Comment by jwhui** [comment]
I have yet to see the PAN ID Query test fail on other travis builds.  Does the failure occur on every build in your branch?

**Comment by nibanks** [comment]
It seems to happen at random. My refactor and master branches are exact copies and I have seen several instances where refactor succeeds and master fails that test.

**Comment by jwhui** [comment]
Have you seen this failure occur using the same code as openthread/openthread master?

**Comment by nibanks** [comment]
I don't ever have any branches using just plain openthread/master code.

**Comment by nibanks** [comment]
Do you guys have any further comments on this PR?

**Comment by jwhui** [comment]
Regarding the solution file, I agree with @darconeous that it would be nice not to have this in the root directory.  I noticed [mbedtls](https://github.com/ARMmbed/mbedtls/tree/development/visualc/VS2010) places the solution file in the same directory as the `.vcxproj` files.

File: src/core/thread/panid_query_server.cpp [hunk 16]
```diff
@@ -100,7 +100,7 @@ ThreadError PanIdQueryServer::SendQueryResponse(const Coap::Header &aRequestHead
                                                 const Ip6::MessageInfo &aRequestInfo)
 {
     ThreadError error = kThreadError_None;
-    Message *message;
+    Message *message = NULL;
```
Comment (nibanks) [inline src/core/thread/panid_query_server.cpp]: Yes, it's a VS warning; for the path it goes through the `assert`.

File: src/core/thread/topology.hpp [hunk 17]
```diff
@@ -73,6 +73,9 @@ class Neighbor
      *
      */
     enum State
+#if _WIN32
```
Comment (nibanks) [inline src/core/thread/topology.hpp]: No, not unless all platforms we build on support this. I added a unit test case to show/validate the problem.

File: third_party/mbedtls/mbedtls-config.h [hunk 18]
```diff
@@ -28,11 +28,18 @@
 #ifndef MBEDTLS_CONFIG_H
 #define MBEDTLS_CONFIG_H
 
-#include <inttypes.h>
 #include <stdlib.h>
 
 #include <platform/logging.h>
 
+#if defined(_WIN32)
```
Comment (nibanks) [inline third_party/mbedtls/mbedtls-config.h]: Not without editing the mbedtls code directly. I don't have permission to edit that project, just OpenThread. So until I get permission, I can't fix these the correct way.

**Comment by nibanks** [comment]
@draconeous mentioned that all files that start with a `.` are hidden on unix systems. What if I renames the solution to `.openthread.sln`?

**Comment by darconeous** [comment]
Files with a `.` prefix aren't shown when you do a `ls` without the `-h` flag. We generally only want to hide files that aren't really a part of the sources for the project (Like `.travis.yml`, `.appveyor.yml`, `.gitignore`, etc.), are under-the-hood parts of the build system (`.default-version`). They are things that we generally don't want people to see because they don't need to see them. Ideally those files would be hidden on Windows, too.

In this case, we *want* people to see `openthread.sln`. We just don't want it in the root folder.

**Comment by nibanks** [comment]
Ok. I will move it.

File: src/core/thread/topology.hpp (hunk 17 above)
Comment (darconeous) [inline src/core/thread/topology.hpp]: We might be able to address this with an `OT_PLAT` preprocessor macro... Perhaps something like `OT_PLAT_ENUM_CONTAINER(unsigned int)`. It would be defined to be `: type` on WIN32 and nothing on other platforms.

File: tests/unit/test_toolchain.cpp [hunk 19]
```diff
@@ -74,11 +75,21 @@ void test_packed_union()
     VerifyOrQuit(sizeof(packed_t) == 5, "Toolchain::OT_TOOL_PACKED failed 3\n");
 }
 
+void test_packed_enum()
+{
+    Thread::Neighbor neighbor;
+    neighbor.mState = Thread::Neighbor::kStateValid;
+
+    // Make sure that when we read the 3 bit field it is read as unsigned, so it return '4'
```
Comment (darconeous) [inline tests/unit/test_toolchain.cpp]: Fascinating. So does this test pass on other platforms?

File: include/openthread-types.h [hunk 20]
```diff
@@ -249,15 +249,18 @@ typedef struct otExtAddress
...
     union
     {
         uint8_t  m8[OT_IP6_ADDRESS_SIZE];                      ///< 8-bit fields
         uint16_t m16[OT_IP6_ADDRESS_SIZE / sizeof(uint16_t)];  ///< 16-bit fields
         uint32_t m32[OT_IP6_ADDRESS_SIZE / sizeof(uint32_t)];  ///< 32-bit fields
     } mFields;                                                 ///< IPv6 accessor fields
-} OT_TOOL_PACKED_END otIp6Address;
+} OT_TOOL_PACKED_END;
+
+typedef struct otIp6Address otIp6Address;
```
Comment (darconeous) [inline include/openthread-types.h]: 👍

File: src/core/openthread.cpp [hunk 21]
```diff
@@ -885,11 +885,15 @@ void otSetStateChangedCallback(otInstance *, otStateChangedCallback aCallback, v
 const char *otGetVersionString(void)
 {
     static const char sVersion[] =
-        PACKAGE_NAME "/" PACKAGE_VERSION "; "
+        PACKAGE_NAME "/" PACKAGE_VERSION
 #ifdef  PLATFORM_INFO
-        PLATFORM_INFO "; "
+        "; " PLATFORM_INFO
+#endif
+#if defined(__DATE__)
```
Comment (darconeous) [inline src/core/openthread.cpp]: 👍

File: tests/unit/test_toolchain.cpp (hunk 19 above)
Comment (nibanks) [inline tests/unit/test_toolchain.cpp]: On Windows this fails without the other change to the State definition.
